mlabs/src/Mlabs/EfficientNFT/Token.hs

```diff
@@ -136,31 +109,42 @@ mkPolicy oref authorPkh royalty platformConfig _ mintAct ctx =
           mpShare = fromEnum $ pcMarketplaceShare platformConfig
 
           authorAddr = pubKeyHashAddress authorPkh
-          authorShare = Ada.lovelaceValueOf $ price' * 10000 `div` royalty'
+          authorShare = Ada.lovelaceValueOf $ price' * 10000 `divide` royalty'
 
           marketplAddr = pubKeyHashAddress (pcMarketplacePkh platformConfig)
-          marketplShare = Ada.lovelaceValueOf $ price' * 10000 `div` mpShare
+          marketplShare = Ada.lovelaceValueOf $ price' * 10000 `divide` mpShare
 
-          !curSymDatum = Datum $ PlutusTx.toBuiltinData $ ownCurrencySymbol ctx
-          !datums = txInfoData info
+          curSymDatum = Datum $ PlutusTx.toBuiltinData $ ownCurrencySymbol ctx
 
           checkPaymentTxOut addr val (TxOut addr' val' dh) =
-            addr == addr' && val == val' && (dh >>= (`lookup` datums)) == Just curSymDatum
-       in any (checkPaymentTxOut authorAddr authorShare) outs
+            addr == addr' && val == val'
+            && (dh >>= \dh' -> findDatum dh' info) == Just curSymDatum
+       in 
+          any (checkPaymentTxOut authorAddr authorShare) outs
             && any (checkPaymentTxOut marketplAddr marketplShare) outs
-
+            
+-- todo: docs
 {-# INLINEABLE mkTokenName #-}
 mkTokenName :: PubKeyHash -> Natural -> TokenName
 mkTokenName (PubKeyHash pkh) price =
-  TokenName $ sha2_256 $ pkh <> toBuiltin (toStrict (Binary.encode (fromEnum price)))
+  TokenName $ sha2_256 $ (pkh <> toBin (fromEnum price))
+
+{-# INLINEABLE toBin #-}
+toBin :: Integer -> BuiltinByteString
```

Comment:
we can keep this safe but inefficient way of doing this for now until we find a better way